### DISCUSSION TITLE
Allow packageRules allowedVersions to be set multiple times

### DISCUSSION THREAD

**Discussion by Viserius** [renovatebot/renovate#21213] [Request Help]
### What would you like Renovate to be able to do?

I would like Renovate to have the ability to apply the allowedVersions property multiple times for different identifiers that match the same package, and have them all evaluated by Renovate.

Here's an example renovate.json file:
```json
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "extends": [
    "local>swh-dev/basisvoorzieningen/ssd/renovate/presets:base",
    "local>swh-dev/basisvoorzieningen/ssd/renovate/presets:registries"
  ],
  "packageRules": [
    {
      "matchPackageNames": ["jakarta.activation:jakarta.activation-api"],
      "allowedVersions": "<2.0.0"
    },
    {
        "managers": ["maven"],
        "allowedVersions": "!/redhat/"
    },
  ]
}
```

Suppose there is a pom.xml file that contains the following dependency:
```
<dependency>
                <groupId>jakarta.activation</groupId>
                <artifactId>jakarta.activation-api</artifactId>
                <version>1.2.2</version>
            </dependency>
```

New versions found by renovate include: 2.1.1 and 1.2.2.redhat-00001

After running Renovate, new versions (2.1.1 and 1.2.2.redhat-00001) are found. According to the provided JSON, Renovate should not allow any versions that include 'redhat', and only versions below 2.0.0 for the specified 'jakarta.activation-api' package. However, despite the multiple rules specified, Renovate seems to overwrite the first rule with the second.

### If you have any ideas on how this should be implemented, please tell us here.

To improve Renovate's functionality, all rules should be applied one-by-one when considering if a new version for a package matches the specified ruleset, rather than only considering the last rule. This is especially important for the allowedVersions rule, which can be quite restrictive. With this change, users would be able to define more general rules that span a wider range of packages, as well as more specific rules that apply to individual packages or smaller subsets.

### Is this a feature you are interested in implementing yourself?

No

**Comment by secustor** [comment]
Renovate applies packageRules one after another.     
As both rules matches, the first is applied following by the second. Because of this, `allowedVersions` is overwritten by the second rule. 

You can change the order of rules or use an exclude matchers such as [excludePackageNames](https://docs.renovatebot.com/configuration-options/#excludepackagenames)

**Reply by Viserius** [reply]
I am aware of how it works right now, this is also what I write in my original post. We argue it makes more sense to apply all applications of allowedVersions, and not just the last.

**Reply by secustor** [reply]
It is unlikely we will change the merge behavior as this would be very confusing for user if presets contain the same config options as renovate.json . 

What I can image is to add an additional negate option for `allowedVersions` e.g. `forbiddenVersions`. 
This would result in such packageRules: 

```
  "packageRules": [
    {
      "matchPackageNames": ["jakarta.activation:jakarta.activation-api"],
      "allowedVersions": "<2.0.0"
    },
    {
        "managers": ["maven"],
        "forbiddenVersions": "/redhat/"
    },
```

Would this still fit your use case?

**Reply by viceice** [reply]
maybe the forbidden version can be a mergable array and the version shouldn't match any of them?

**Reply by viceice** [reply]
@rarkins @secustor @JamieMagee WDYT?

**Reply by rarkins** [reply]
We could make `allowedVersions` an array, but then would that array be AND or OR? It sounds like it would need to be AND in order to satisfy this particular request, but does that make sense for all?

**Reply by secustor** [reply]
It should be OR to be consisted with other config options such as `autodiscoverFilter` and matchers. 

I let it sink a little and AND is more useful for users even if we are not consistent with other options.

**Reply by Viserius** [reply]
If I understand the proposal of @secustor correctly, adding a forbiddenVersions-tag would be equivalent to adding the entered value to allowedVersions but negate it. Yet, this would work in our case, simply because we have 2 properties we can use for restricting versions, such that we can use the allowedVersions on a specific dependency and forbiddenVersions on a global level. 

If this would be a mergeable array like @viceice comments, I think you have the most flexibility regarding building a tier-structure. For instance, you can set a forbiddenVersion on `-M\d` (for milestones) for all packages, another restriction for maven packages (such as the 'redhat' term), another restriction for everything in the jakarta package group, and maybe another restriction on an exact match on a specific dependency.

@rarkins To me it makes sense that if you define a version whitelist on a group (>1.0.0) and a rule for a more specific scope (>2.0.0), you would apply both, analogous to an AND. Yet, then you no longer have the ability to overwrite the initial whitelist. E.g., >1.0.0 for a group, and >0.0.1 & <1.0.0 for a narrower scope, which you are able to do now (not that we apply this). The case could be made whether the former/wider whitelist was scoped appropriately and perhaps should have been more specific, but that is for you to decide :). 

Either way, merging allowedVersions using `AND` or adding a forbiddenVersions option would both work for us, and would be extremely helpful for us to specificy some global organization-wide restrictions of packages we do not want to pop up, while still being able to apply restrictions on a more narrow level. Thank you for your efforts and thinking along so far.

**Reply by rarkins** [reply]
Changing the behavior of `allowedVersions` in a way which stops you being able to override it would be a breaking change. The question is: would anyone still have the need for the old functionality, where "last rule wins"? If the answer is yes, then we shouldn't change `allowedVersions`, which I guess means we'd have to consider a new field which has the "AND" semantic:
- it's mergeable (each item is an AND requirement)
- it's ANDed together with `allowedVersions`

**Reply by Viserius** [reply]
@rarkins Do you know if any concrete changes have been made for this already?

**Reply by rarkins** [reply]
None

**Reply by tarioch** [reply]
Just to confirm, there is still nothing to solve this case at the moment, correct? Running into the same issue (also with the redhat versions).
